File: scripts/apply_migration.py
#!/usr/bin/env python3
"""
Apply Phase 1 security migration to Supabase database.
This script executes the SQL migration using the Supabase REST API.
"""

import os
import sys
from pathlib import Path
from dotenv import load_dotenv
from supabase import create_client, Client

# Load environment variables from .env.local
env_path = Path(__file__).parent.parent / '.env.local'
load_dotenv(dotenv_path=env_path)

def apply_migration():
    """Apply the Phase 1 security migration."""
    print("📦 Loading Supabase credentials...")
    
    supabase_url = os.getenv('NEXT_PUBLIC_SUPABASE_URL')
    supabase_key = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
    
    if not supabase_url or not supabase_key:
        print("❌ Error: Missing Supabase credentials in .env.local")
        sys.exit(1)
    
    print(f"✅ Connected to: {supabase_url}\n")
    
    # Create Supabase client
    supabase: Client = create_client(supabase_url, supabase_key)
    
    # Read migration file
    migration_path = Path(__file__).parent.parent / 'supabase' / 'migrations' / '02_phase1_security.sql'
    print(f"📄 Reading migration: {migration_path.name}")
    
    with open(migration_path, 'r', encoding='utf-8') as f:
        sql = f.read()
    
    print("\n🚀 Applying Phase 1 migration to database...")
    print("   This will create:")
    print("   - user_sessions table (session tracking)")
    print("   - trusted_devices table (device management)")
    print("   - audit_log table (tamper-evident logging)")
    print("   - usage_meters table (quota tracking)")
    print("   - Helper functions and RLS policies\n")
    
    try:
        # Execute SQL via Supabase PostgREST
        # Note: Supabase Python client doesn't support direct SQL execution
        # We need to use the SQL Editor in dashboard or use psycopg2
        
        print("⚠️  Note: Supabase Python client doesn't support direct SQL execution.")
        print("\n📝 Please apply the migration manually using one of these methods:\n")
        
        print("METHOD 1: Supabase Dashboard (Recommended)")
        print("=" * 60)
        print("1. Go to: https://supabase.com/dashboard")
        print("2. Select your project")
        print("3. Navigate to: SQL Editor")
        print("4. Click: New Query")
        print("5. Copy and paste the contents of:")
        print(f"   {migration_path}")
        print("6. Click: Run\n")
        
        print("METHOD 2: psql Command Line")
        print("=" * 60)
        print("If you have PostgreSQL client installed:")
        print(f"psql -h tybaetnvnfgniotdfxze.supabase.co -p 5432 -d postgres -U postgres -f {migration_path}")
        print("(You'll be prompted for the database password)\n")
        
        print("METHOD 3: Using psycopg2 (Install first)")
        print("=" * 60)
        print("pip install psycopg2-binary")
        print("Then run: python scripts/apply_migration_psycopg2.py\n")
        
        # Try to verify if tables already exist
        print("🔍 Checking if migration already applied...")
        try:
            # Check if user_sessions table exists
            result = supabase.table('user_sessions').select('id').limit(1).execute()
            print("✅ Migration appears to be already applied (user_sessions table exists)")
            return True
        except Exception as check_error:
            if 'relation' in str(check_error).lower() or 'does not exist' in str(check_error).lower():
                print("❌ Migration not yet applied (tables don't exist)")
                print("\n👉 Please use METHOD 1 (Dashboard) above to apply the migration.")
                return False
            else:
                print(f"⚠️  Could not verify: {check_error}")
                return False
        
    except Exception as e:
        print(f"❌ Error: {e}")
        return False

if __name__ == '__main__':
    success = apply_migration()
    sys.exit(0 if success else 1)

File: scripts/apply_migration.js
const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
const path = require('path');

// Load environment variables
require('dotenv').config({ path: '.env.local' });

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL;
const supabaseServiceKey = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!supabaseUrl || !supabaseServiceKey) {
  console.error('Missing Supabase credentials in .env.local');
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseServiceKey);

async function applyMigration() {
  console.log('📦 Reading migration file...');
  
  const migrationPath = path.join(__dirname, '..', 'supabase', 'migrations', '02_phase1_security.sql');
  const sql = fs.readFileSync(migrationPath, 'utf8');
  
  console.log('🚀 Applying Phase 1 migration to database...');
  console.log('   - Creating user_sessions table');
  console.log('   - Creating trusted_devices table');
  console.log('   - Creating audit_log table');
  console.log('   - Creating usage_meters table');
  console.log('   - Adding role_type to user_profiles');
  console.log('   - Creating helper functions\n');
  
  try {
    // Execute the migration SQL
    const { data, error } = await supabase.rpc('exec_sql', { sql_query: sql });
    
    if (error) {
      // If rpc doesn't exist, try direct execution (Supabase might not support this)
      console.log('⚠️  RPC method not available, trying alternative approach...\n');
      
      // Split SQL into individual statements and execute
      const statements = sql
        .split(';')
        .map(s => s.trim())
        .filter(s => s.length > 0 && !s.startsWith('--'));
      
      for (let i = 0; i < statements.length; i++) {
        const stmt = statements[i];
        if (stmt) {
          console.log(`   Executing statement ${i + 1}/${statements.length}...`);
          const { error: execError } = await supabase.from('_').select('*').limit(0); // This won't work
          
          if (execError) {
            console.error(`   ❌ Error on statement ${i + 1}:`, execError.message);
          }
        }
      }
    } else {
      console.log('✅ Migration applied successfully!\n');
      console.log('📊 New tables created:');
      console.log('   - user_sessions (session tracking)');
      console.log('   - trusted_devices (device management)');
      console.log('   - audit_log (tamper-evident logging)');
      console.log('   - usage_meters (quota tracking)\n');
      console.log('🔧 New functions created:');
      console.log('   - get_active_sessions_count()');
      console.log('   - get_current_month_usage()');
      console.log('   - increment_usage_meter()');
      console.log('   - create_audit_log_entry()');
      console.log('   - cleanup_expired_sessions()\n');
    }
  } catch (err) {
    console.error('❌ Migration failed:', err.message);
    console.error('\n📝 Please apply the migration manually:');
    console.error('   1. Go to https://supabase.com/dashboard');
    console.error('   2. Select your project');
    console.error('   3. Go to SQL Editor');
    console.error('   4. Copy the contents of: supabase/migrations/02_phase1_security.sql');
    console.error('   5. Paste and run\n');
    process.exit(1);
  }
}

applyMigration();
